x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/map/LogWithBaseInMap.java

```diff
@@ -0,0 +1,209 @@
+        );
+    }
+
+    @Override
+    public final void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(number);
+        out.writeOptionalNamedWriteable(map);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        // validate field type
+        TypeResolution resolution = isNumeric(number, sourceText(), FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        if (map != null) {
+            // MapExpression does not have a DataType associated with it
+            resolution = isMapExpression(map, sourceText(), SECOND);
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+            return validateOptions();
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DOUBLE;
+    }
+
+    @Override
+    public boolean foldable() {
+        return number.foldable();
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new LogWithBaseInMap(source(), newChildren.get(0), newChildren.size() > 1 ? newChildren.get(1) : null);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, LogWithBaseInMap::new, number, map);
+    }
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
+        var valueEval = Cast.cast(source(), number.dataType(), DataType.DOUBLE, toEvaluator.apply(number));
+        double base = Math.E;
+        if (map instanceof MapExpression me) {
+            Expression b = me.get(BASE);
+            if (b != null && b.foldable()) {
+                Object v = b.fold();
+                if (v instanceof BytesRef br) {
+                    v = br.utf8ToString();
+                }
+                base = Double.parseDouble(v.toString());
+            }
+        }
+        return new LogWithBaseInMapEvaluator.Factory(source(), valueEval, base);
+    }
+
+    @Evaluator(warnExceptions = { ArithmeticException.class })
+    static double process(double value, @Fixed double base) throws ArithmeticException {
+        if (base <= 0d || value <= 0d) {
+            throw new ArithmeticException("Log of non-positive number");
+        }
+        if (base == 1d) {
+            throw new ArithmeticException("Log of base 1");
+        }
+        return Math.log10(value) / Math.log10(base);
+    }
+
+    public Expression number() {
+        return number;
+    }
+
+    public Expression base() {
+        return map;
+    }
+
+    private TypeResolution validateOptions() {
```

Comment:
It's not needed for this PR, but we will need to provide a validation mechanism for map options so they are checked for valid keys, valid types etc